### PR TITLE
#8 proximity based timer

### PR DESCRIPTION
**Pull request by Einfachmario45**
#8

### CODE REVIEW

**Comment by LeonUwe** [comment]
The hack-button is never made invisible again. Once a driver has completed the proximity timer, they'll have the button for the rest of their time playing. It should disappear when they leave the proximity of the hackable vehicle.

**Comment by LeonUwe** [comment]
There is a lot of reformatting going on in the .html and .css files. This makes reviewing them quite difficult. In the future please isolate the reformatting in its own commits or refrain from doing it completely.

**Comment by LeonUwe** [comment]
Apart from the mentioned issues, the functionality has been implemented nicely.

**Comment by LeonUwe** [comment]
I found a little bug:

image

This was caused by kicking a player out of the race (manually or by being idle). This message is printed to the console repeatedly until the player reconnects.

Should be fixable by a is None check in DriverUI.py line 258.

**Comment by Einfachmario45** [comment]
> 

Sorry about the reformatting. I have Prettier installed, and it reformats automatically.

**Review by LeonUwe — CHANGES_REQUESTED** [review]

File: src/UserInterface/CarMap.py [hunk 1]
```diff
@@ -88,6 +90,24 @@ def update_virtual_location(self, vehicle_id: str, position: dict, angle: float)
...
+        if proximity_vehicle_id != None:
+            pos_proximity_vehicle = self._environment_manager.get_vehicle_by_vehicle_id(proximity_vehicle_id)._location_service._current_position
+            if pos_proximity_vehicle.distance_to(pos_self) > 200:
+                self._environment_manager.get_vehicle_by_vehicle_id(vehicle_id).vehicle_in_proximity = None
+        else:
+            for vehicle in self._vehicles:
+                if vehicle.vehicle_id != vehicle_id:
+                    pos_other =vehicle._location_service._current_position
+                    if pos_other.distance_to(pos_self) < 200:
+                        self._environment_manager.get_vehicle_by_vehicle_id(vehicle_id).vehicle_in_proximity = vehicle.vehicle_id
```
[the remark this answers is not in the document]
Comment (LeonUwe) [reply src/UserInterface/CarMap.py:121]: Fixed in [833673a](https://github.com/LeonUwe/IAV_Distortion/pull/21/commits/833673a6a132f58e4fd6fc6709917b12da0a8d0d)

File: src/UserInterface/CarMap.py [hunk 2]
```diff
@@ -88,6 +90,24 @@ def update_virtual_location(self, vehicle_id: str, position: dict, angle: float)
         """
         data = {'car': vehicle_id, 'position': position, 'angle': angle}
         self.__run_async_task(self.send_car_position(data))
+        
+        if self._vehicles is not None:
+            self.check_vehicle_proximity(vehicle_id,position)             
+        return
+
+    def check_vehicle_proximity(self,vehicle_id: str, position: dict,) -> None:
```
[the remark this answers is not in the document]
Comment (LeonUwe) [reply src/UserInterface/CarMap.py:98]: [833673a](https://github.com/LeonUwe/IAV_Distortion/pull/21/commits/833673a6a132f58e4fd6fc6709917b12da0a8d0d)

File: src/UserInterface/DriverUI.py [hunk 3]
```diff
@@ -229,6 +238,27 @@ async def __check_driver_heartbeat_timeout(self):
                     logger.info(f'Player {player} timed out. Removing player from the game...')
                     self.__remove_player(player)
 
+    async def __send_proximity_vehicle(self, player: str):
```
[the remark this answers is not in the document]
Comment (LeonUwe) [reply src/UserInterface/DriverUI.py:241]: [833673a](https://github.com/LeonUwe/IAV_Distortion/pull/21/commits/833673a6a132f58e4fd6fc6709917b12da0a8d0d)

File: src/UserInterface/DriverUI.py [hunk 4]
```diff
@@ -229,6 +238,27 @@ async def __check_driver_heartbeat_timeout(self):
...
+                if previous_value != vehicle.vehicle_in_proximity:
+                    uuid = vehicle.vehicle_id
+                    await self._sio.emit('send_proximity_vehicle',{'vehicle_id': vehicle.vehicle_in_proximity,'proximity_timer': self.__driver_proximity_timer})
+                    previous_value= vehicle.vehicle_in_proximity
+                    vehicle.proximity_timer = time.time()
+                else:
+                    if vehicle.vehicle_in_proximity != None:
+                        self.__run_async_task(self.__check_driver_proximity_timer(player))
+
+    async def __check_driver_proximity_timer(self, player: str):
```
[the remark this answers is not in the document]
Comment (LeonUwe) [reply src/UserInterface/DriverUI.py:263]: [833673a](https://github.com/LeonUwe/IAV_Distortion/pull/21/commits/833673a6a132f58e4fd6fc6709917b12da0a8d0d)

File: src/UserInterface/templates/driver_index.html [hunk 5]
```diff
@@ -1,210 +1,316 @@
...
-                socket.emit('driver_inactive', {player: '{{player}}'});
-            }
-            else {
-            // The page is visible again
-                socket.emit('driver_active', {player: '{{player}}'});
-            }
-        });
+      socket.on('send_finished_proximity_timer', function(vehicle_id){
+        document.getElementById("hack-button").style.display = "block";
+      });
```
[the remark this answers is not in the document]
Comment (LeonUwe) [reply src/UserInterface/templates/driver_index.html]: [833673a](https://github.com/LeonUwe/IAV_Distortion/pull/21/commits/833673a6a132f58e4fd6fc6709917b12da0a8d0d)

**Review by LeonUwe — APPROVED** [review]

**Pull request merged**